src/java/org/apache/cassandra/io/tries/TrieNode.java

```diff
@@ -0,0 +1,968 @@
+        {
+            int childCount = node.childCount();
+            assert childCount > 0;
+            assert childCount < 256;
+            dest.writeByte((ordinal << 4) + (payloadBits & 0x0F));
+            dest.writeByte(childCount);
+
+            for (int i = 0; i < childCount; ++i)
+                dest.writeByte(node.transition(i));
+            for (int i = 0; i < childCount; ++i)
+                writeBytes(dest, -node.serializedPositionDelta(i, nodePosition));
+        }
+    };
+
+    static final Sparse12 SPARSE_12 = new Sparse12();
+    static class Sparse12 extends Sparse
+    {
+        // byte flags
+        // byte count (<= 255)
+        // count bytes transitions
+        // count 12-bits transition targets
+        // var payload
+        Sparse12()
+        {
+            super(FRACTIONAL_BYTES);
+        }
+
+        @Override
+        public int payloadPosition(ByteBuffer src, int position)
+        {
+            int count = transitionRange(src, position);
+            return position + 2 + (5 * count + 1) / 2;
+        }
+
+        @Override
+        public long transitionDelta(ByteBuffer src, int position, int childIndex)
+        {
+            return -read12Bits(src, position + 2 + transitionRange(src, position), childIndex);
+        }
+
+        @Override
+        public int sizeofNode(SerializationNode<?> node)
+        {
+            return 2 + (node.childCount() * 5 + 1) / 2;
+        }
+
+        @Override
+        public void serialize(DataOutput dest, SerializationNode<?> node, int payloadBits, long nodePosition) throws IOException
+        {
+            int childCount = node.childCount();
+            assert childCount < 256;
+            dest.writeByte((ordinal << 4) + (payloadBits & 0x0F));
+            dest.writeByte(childCount);
+
+            for (int i = 0; i < childCount; ++i)
+                dest.writeByte(node.transition(i));
+            int i;
+            for (i = 0; i + 2 <= childCount; i += 2)
+            {
+                int p0 = (int) -node.serializedPositionDelta(i, nodePosition);
+                int p1 = (int) -node.serializedPositionDelta(i + 1, nodePosition);
+                assert p0 > 0 && p0 < (1 << 12);
+                assert p1 > 0 && p1 < (1 << 12);
+                dest.writeByte(p0 >> 4);
+                dest.writeByte((p0 << 4) | (p1 >> 8));
+                dest.writeByte(p1);
+            }
+            if (i < childCount)
+            {
+                long pd = -node.serializedPositionDelta(i, nodePosition);
+                assert pd > 0 && pd < (1 << 12);
+                dest.writeShort((short) (pd << 4));
+            }
+        }
+
+        @Override
+        boolean fits(long delta)
+        {
+            return 0 <= delta && delta <= 0xFFF;
+        }
+    };
+
+    static final Dense DENSE_16 = new Dense(2);
+    static final Dense DENSE_24 = new Dense(3);
+    static final Dense DENSE_32 = new Dense(4);
+    static final Dense DENSE_40 = new Dense(5);
```

Comment:
nit: Should probably move these into `Dense` itself to avoid potential classloading deadlocks